### DISCUSSION TITLE
Allow users to set "Hide whitespace changes" by default

### DISCUSSION THREAD

**Discussion by ovflowd** [community/community#5486] [Pull Requests] [answered]
This is a feature request that exists for 3 years (https://github.community/t/feature-request-option-to-always-hide-whitespace-changes-when-viewing-diffs/1896) and would be a very nice addition for GitHub.

### The problem

Many users want to by default **always** "Hide whitespace changes" on Git PR diffs or GitHub code diffs in general.

### The solution

- Allow a user-wide setting to "Hide whitespace changes" (And vice-versa, as a checkbox)

**Comment by golaod** [comment]
If not as a setting, then as part of a GET param or a cookie.

**Reply by zeitwidrig** [reply]
GET param already exists. Add `?w=1` to ignore whitespace changes.

**Reply by rbelling** [reply]
that works @zeitwidrig but you have to add it every time and it seems like it gets cleared when you navigate to another GH page

**Comment by WaKeMaTTa** [comment]
Sometimes I move code and the PR that I make seems very complex, but when you use "Hide whitespace changes" you can easliy realize that one change a couple of lines.

I would love to be able to configure this for each PR differently.

**Reply by EskeRahn** [reply]
I would claim that it would be better to have individual settings for file types, for the project.

As e.g. indentation is important for _some_ file types.
Strictly it is the change in indentations that matters here, so if a large block is moved in/out, only the beginning and end of the block is interesting. For many file types we might want to completely ignore repetition of whitespaces.

**Comment by willsmythe** [comment, marked as answer]
This is partially addressed by a feature we shipped a few weeks ago: https://github.blog/changelog/2021-10-14-hiding-whitespace-is-now-remembered-for-each-pull-request/

GitHub now remembers "hide whitespace" for you on the pull requests you enable it on. 

We explored other options including a user-level setting (and we may reassess in the future), but decided the right first step was to remember it for each pull request.

**Reply by rbelling** [reply]
thanks @willsmythe this is a great first step. Also I like how that setting is more visible overall in the UI

**Reply by ovflowd** [reply]
Thanks! This little change is really appreciate 😄

**Reply by osdiab** [reply]
Any chance you might still do the user-level choice (or even repo-level)? My use case is that we use a code formatter (Prettier) that enforces consistent whitespace already, so whitespace is almost always a (very disruptive) distraction from the substance of our PRs (a 1 line change can become a 50 line change if nesting is edited, for example). By PR is only marginally useful, because after an initial PR review, I won't be looking at that PR very much again, and the underlying reason why whitespace changes are so uninteresting is a function of the properties of this particular project, lending me to think that a project- or project + user combo-mediated default is the most sensical for ours and very many other repos.

Hoping to see this change happen sometime soon! (Though more imminently important is that ignoring white space comes to the mobile UI at all).

**Reply by chelmertz** [reply]
@willsmythe  @michellemerrill What is the rationality behind marking the "first step", which "partially addresses" the feature request, as an answer? (Your/Github's words in quotes.)

All links in different forums points to this canonical issue, so when asking for this feature elsewhere, those discussions are dead in the water and I guess (since this is, again, "marked as answer"), this issue is excluded from Github's todo-list/prioritization.

Doing this in an iterative way could have been to create a separate sub-discussion, solve that and link it from here but... here you get me nagging away instead :)

**Reply by lpar** [reply]
This impacts every Go project, as Go's standard formatter (which pretty much everyone uses) enforces consistent whitespace.

**Reply by archon810** [reply]
For those who go through a lot of PRs, most of the time merge them without requesting changes, the per-PR setting is not nearly as helpful as a repo-level or user-level setting. Please add those.

**Reply by BurkhardWick** [reply]
I'm usually only opening each PR once, so this is completely useless.

**Reply by thatguyintech** [reply]
+100 I'd really love a user-level or repo-level setting, I pretty much review every single PR with "remove whitespace" turned on

**Reply by CecilePalant** [reply]
This is absolutely far from the answer to this thread. Please un-mark it the answer.
I review a lot of PRs and only go there once, having to click navigate and make 3 clicks to 'hide whitespace' on each and every single PR is not solved by that 'partial addressing'.

**Reply by cupcakearmy** [reply]
It neeeeds to be a user setting @willsmythe There are literaly chrome/firefox extensions that exists for that purpose. Seems such an easy win for GH in theory. 
This is drastically needed! ❤️

**Reply by rusterholz** [reply]
Agreed, this is not an actual answer to the question. Indeed, it openly admits that GitHub has not actually addressed the option being requested. @willsmythe please unmark this as the answer and removed the `shipped` label, as they are both actively misleading.

**Reply by joe-lloyd** [reply]
### I made a plugin

I had this issue so i made a Google Chrome plugin to fix this. You can find it [here](https://chrome.google.com/webstore/detail/github-whitespace-hider/fapgonkhjoinbbcachbpcjhheikgbjem/)

Also for full transparency go [here](https://github.com/joe-lloyd/github-white-space-off-chrome-extention) to see the repo where the code is.

Marked as not trusted by google since its my first time making a plugin. Apparently it takes a couple of months to become trusted.

**Reply by q-ahmedsalman** [reply]
This is not an answer.

**Reply by aiguofer** [reply]
Kinda crazy this isn't a setting I can choose FOR ALL PRs. Also insane that this would be considered an answer.

The more I do stuff on Github the more I miss Gitlab... I sure miss all the options and configurations I had there. It's so annoying to have to enable this option every single time I visit some random PR for either work or some open source project.

**Reply by EskeRahn** [reply]
Dear @willsmythe , the common issue for small PRs are _not_ that they are revisited many times by each user.
The common scenario is many many small PRs only visited once or more rarely twice by a reviewer.  

It would be SO much better if the repo got a common flag reflecting the code style used in the project, or more advanced a set of flags matching the file-types. e.g. we DO want yaml-files to care about indentation, but we do not care about whitespaces in e.g. C# files.  

The saved flag _will_ help for huge PRs that take a long time and many re-visits to review, but I would claim this a special case, as the huge PR often could have benefitted from being broken down to several smaller in the first place. I know this is not _always_ possible though, and for these special cases the saved flag can help. But it is not an answer to the suggestion

...And if huge, it is also likely to contain multiple file types, where the relevant setting of the flag might not be the same...

**Reply by sholsinger** [reply]
I agree that per-user-and-repo would be an MVP on this.

We already have the ability to store a preference for which email should be used for each repo. Can we not also add a flag for whitespace to that relationship?

**Reply by colinphill-mdsol** [reply]
@willsmythe @michellemerrill It's great that a forward step has been taken, but it is not equivalent to the feature being requested here, so should this question really be marked as answered and shipped? This is still a valid outstanding feature request.

**Reply by chicken-suop** [reply]
This question is not answered. I would like this feature in 2024

**Reply by brett-adim** [reply]
Partially addressed !== addressed. This doesn't reduce the amount of work it takes to do a proper review.

**Reply by TD-DO** [reply]
Yes please implement this feature all the way

**Reply by johnc-ftl** [reply]
@willsmythe - This is not remotely close to addressing the issue. Even a per PR setting would be much better (b/c many PRs are hard to review without telling every participant to click a link with `&w=1`), but really this needs to be per repo.

**Reply by xbtmatt** [reply]
Could you please add this? I'd wager some 95%+ of diffs have 5x the amount of lines changed due to the indentation changing...

**Reply by paul-titlevision** [reply]
I'd like to be able to select Hide Whitespaces once and never have to do it again.

**Reply by RobertTheBlair** [reply]
Just  about    every    day, I waste  time in my  PRs going "Why did  this file change. ?  Oh, duh, it was a IDE generated white space change.   How   I   wish there was a setting to help with this"
Just fix it github,   and the sighs of 20,000 developers will resound  through the  ether.

**Reply by Kaylebor** [reply]
So it's supposed to be something that GitHub remembers? What does it mean?
If it's per-tab, then that's next to useless; with my usage patterns, every single time I visit any PR and look at the file diffs, it is unchecked without fail.

At this point it may be better to just add a Redirector rule on my browser to always append `?w=1` on all GitHub PR URLs, at least that will be consistent.

**Reply by bangonkali** [reply]
It would be great if this can be a per user/account setting. I would need this in all areas where diffing UI is involved PR and Commit Hash Pages.

**Reply by qneill-sifive** [reply]
Everyone is arguing for linters and formatters - without knowing they are arguing for linters and formatters.

**Reply by lpar** [reply]
It's a problem for Go code, and Go has a standard formatter pretty much everyone uses. Making all code fit a standard format doesn't solve the problem.

**Reply by gregmarr** [reply]
@qneill-sifive No, we absolutely are not.  There are plenty of uses that don't get addressed by linters and formatters.  In fact, the existence of linters and formatters makes this need MORE important.  As an example of a use case, take a block of code and put it inside a conditional block, so your linter and formatter indents that code.  Now all of those lines of code have whitespace-only differences.  Without "ignore whitespace", then all of those whitespace-only changes are shown, and can cause the difference engine to mis-align those lines.  With "ignore whitespace" on, then the diff shows only the addition of the conditional, and you can easily see that the rest of the lines have no significant changes (ignoring languages where leading whitespace is significant).

**Comment by aidantorrence** [comment]
Any updates?

**Comment by Pakon2543** [comment]
https://github.blog/changelog/2021-10-14-hiding-whitespace-is-now-remembered-for-each-pull-request/

**Comment by sethfri** [comment]
Hi @willsmythe, could you remove the `shipped` label from this discussion? The feature request here is for a _per-user_ setting, applied to all PRs for that user. Are there any plans for this?

**Reply by fiddur** [reply]
It says @michellemerrill marked it as the answer; maybe they could remove that...

**Comment by vtopc** [comment]
I'm using https://chrome.google.com/webstore/detail/github-whitespace/fnpkdafamnbjoldglihkjjdicofghccm as a temporary fix for that.

**Reply by archon810** [reply]
The only problem is it doesn't let you view diffs with white space enabled anymore.

**Reply by cduflo** [reply]
https://chrome.google.com/webstore/detail/git-well-soon/ehpeaofieafibmhiagianfjjblpnmbdo allows a user override to view whitespace. Also works for github enterprise `git.***.com`

**Reply by serg06** [reply]
Thanks for the link! However it won't work for people whose companies block third-party extensions.

**Reply by rspringAya** [reply]
For the love of all that is good, thank you for that extension link! I've wasted so much time sorting through PRs when I forget to check "hide whitespace"... 

Please please please GitHub, revisit this as a user setting!

**Reply by knodi** [reply]
It's mind-boggling that someone would let it persist per-PR and think they had done what was asked for.

Thank you so much for that extension!

**Comment by osdiab** [comment]
This seems to work for Firefox.
https://addons.mozilla.org/en-US/firefox/addon/github-whitespace-disabler/?utm_source=addons.mozilla.org&utm_medium=referral&utm_content=search

Omar

**Comment by mcskinner** [comment]
I would also prefer this setting exist at a per-user level. It is almost never useful to see whitespace changes and I have to turn this off consistently.

**Comment by CowDotDev** [comment]
Voicing support for a user-level setting to keep Hide Whitespace... it's nice to stay on the PR, sure, but working in a project where whitespace doesn't matter, and is also handled via Prettier/Eslint, this is really annoying to have to toggle it for every single PR.

Any are acceptable options in my mind:
- Remember at User Setting Level
- Remember at Repo Setting Level
- Remember at a Repo Setting Level, by User.

**Comment by cduflo** [comment]
My chrome extension solves for this problem until github decides to support this function: https://chrome.google.com/webstore/detail/git-well-soon/ehpeaofieafibmhiagianfjjblpnmbdo

**Reply by CowDotDev** [reply]
**Note to any employees at GitHub that see this comment...**

This is not a fix, having to download a random extension is a security risk - not trying to accuse the developer, just is a concern.
Clearly there is a continuous stream of vocal support for this type of update, I continue to urge you to reassess this ticket as unanswered. @willsmythe

**Reply by arjun-menon** [reply]
The permissions requested are too broad on this extension. Definitely not installing as it asks for permissions to read all websites, as well as browsing history.

**Reply by cduflo** [reply]
@arjun-menon it's been awhile since your comment, but I did go back to refactor the extension to require less permissions. Thanks for your feedback and apologies for the long delay!

**Reply by arjun-menon** [reply]
@cduflo I'm kind of just waiting for https://github.com/orgs/community/discussions/65597 to be implemented. Also, I use Firefox exclusively now, so I can't install Chrome extensions...

**Comment by Luc45** [comment]
@willsmythe are you aware of the negative feedback of the proposed solution? JFYI

**Comment by jlevitt** [comment]
Would love this to be a user setting.

**Comment by RenatoCoelho-BiggerPockets** [comment]
My team is still interested in this feature, and the "answer" doesn't address the issue.

**Comment by ughitsaaron** [comment]
This would honestly be so helpful and convenient. And as the marked comment states with the world "partially", this isn't, in fact, resolved.

**Comment by kkarthikeyan** [comment]
@willsmythe - not sure why this is marked as `shipped` while the issue remains. And you seem to be unmoved by all previous users comments to remove the `answered` and `shipped` labels. This seems very unresponsive on GitHub's part.

**Reply by osdiab** [reply]
Should just make a new issue/discussion, they’re probably just ignoring this since it says it’s resolved

**Comment by stravant** [comment]
I can't believe it's 2023 and this still isn't possible. Please implement it.

**Comment by EskeRahn** [comment]
This is a problem especially hitting C# (and similar) when using the quite common 'convention' of moving the  
{  
to the beginning of next line, as compare currently sees it as an important 100% synchronisation hit if the indentation matches.  

If they, when syncing the two, gave a sole '{' (and '}' ) with _any_ indentation the same (low) synchronisation weight, it will be much more likely to get the syncing right, and the urgent need for the feature they seem to refuse to implement would be almost gone.

In a more advanced version it could (for syncing purposes) consider a set of pure punctuation chars (braces, parentheses, comma... ) with _any_ indentation as something logically 'belonging' to the and of the previous line. It would make it much more likely to get the syncing right.

Obviously it shall be _displayed_ as in the file, I'm merely talking about the logic for syncing.

**Comment by daveatweaviate** [comment]
remember-the-setting-when-revisiting isn't a solution. Please allow a default 'hide whitespace' setting. If hide-reveal whitespace is already possible, why is adding a config setting hard?

**Reply by rbelling** [reply]
I assume because it requires changes to the account page to add that option and there currently isn't a good section for it, it would probably be something like "Your Code Review Preferences" but there isn't such a section yet. And it has to be applied to each PR you're looking at. And potentially be overwritten by the Pr-specific settings. 

So, not trivial, and also probably not a big enough product priority given that we can already set it on a PR manually.

@willsmythe this would be a good one to bundle in a "Your Code Review Preferences" page when there is enough preferences for that account page to make sense. Any idea if this is in the roadmap?

**Reply by hikeeba** [reply]
Adding a toggle option is trivial, and GitHub isn't known for consistent or amazing UI design, so that shouldn't stop them. But it wouldn't suprise me if you're right that they can't figure out how to make this fit easily...

**Reply by RobertTheBlair** [reply]
My company just switched from Bitbucket to Github, and this is one of **many** quality of life features I am missing.

**Comment by daveatweaviate** [comment]
Make it exactly the same as split or unified view. That persists from PR to PR and it is already next to the per PR whitespace toggle. All that would change is to add whatever persistence they already use for the view to the whitespace toggle.

**Comment by szalapski** [comment]
I must agree.  This seems like a no-brainer to make into a user setting.  Please do so!

**Comment by alexander-dermicus** [comment]
A new discussion was opened on this topic: https://github.com/orgs/community/discussions/65597
Since this one is marked as shipped and answered, let's upvote that one instead. Maybe we get a response then.

**Comment by Sachinthana-Lokuyaddage** [comment]
The "Hide Whitespace Changes" feature is a common functionality found in many code review and version control systems. It allows users to focus on the actual content changes in the code by ignoring whitespace modifications, such as indentation adjustments or space additions/removals.

Discussion:

Focus on Content: When reviewing code changes, developers often want to concentrate on meaningful modifications rather than distracting whitespace adjustments. "Hide Whitespace Changes" enables users to filter out these non-substantive alterations and focus solely on the content modifications.

Improved Readability: Whitespace changes, while important for code style and formatting, might not impact the logic or functionality of the code. By hiding these changes, the code review becomes more readable, emphasizing the essential parts that contribute to the code's functionality.

Reduced Noise: In large codebases or complex changes, whitespace modifications can introduce visual noise, making it harder to identify critical content changes. This feature aids in decluttering the review interface and simplifying the identification of meaningful alterations.

Consistent Presentation: "Hide Whitespace Changes" promotes a consistent presentation of code across different reviews. Developers can maintain a standardized viewing experience, making it easier to identify and understand code modifications consistently.

Enhanced Collaboration: By providing a cleaner and more focused view, this feature can improve collaboration among team members during code reviews. It streamlines the review process, making it more efficient and conducive to constructive feedback on substantive code changes.

In summary, "Hide Whitespace Changes" is a valuable feature that enhances the code review experience by allowing users to filter out non-substantive whitespace modifications. It contributes to improved readability, reduced visual noise, and a more consistent and collaborative code review process.

**Reply by DivaniManarandi** [reply]
thanks

**Comment by Abdullah9990** [comment]
I'm a beginner, what good advice can you give me since you're professionals?
